erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,816 @@
+	memory_term = erts_bld_uint(&hp, NULL, memory);
+        res = MAP2(hp, am_count, count_term, am_memory, memory_term);
+        return res;
+    }
+}
+
+#undef NUM_INFO_MAP_KEYS
+#undef DECL_AM
+
+static Uint
+lookup(HashTable* hash_table, Eterm key)
+{
+    Uint mask = hash_table->mask;
+    Eterm* table = hash_table->term;
+    Uint32 idx = make_internal_hash(key, 0);
+    Eterm term;
+
+    do {
+        idx++;
+        term = table[idx & mask];
+    } while (is_boxed(term) && !EQ(key, (tuple_val(term))[1]));
+    return idx & mask;
+}
+
+static HashTable*
+copy_table(HashTable* old_table, Uint new_size, int rehash)
+{
+    HashTable* new_table;
+    Uint old_size = old_table->allocated;
+    Uint i;
+
+    new_table = (HashTable *) erts_alloc(ERTS_ALC_T_PERSISTENT_TERM,
+                                         sizeof(HashTable) +
+                                         sizeof(Eterm) * (new_size-1));
+    if (old_table->allocated == new_size && !rehash) {
+        /*
+         * Same size and no key deleted. Make an exact copy of the table.
+         */
+        *new_table = *old_table;
+        for (i = 0; i < new_size; i++) {
+            new_table->term[i] = old_table->term[i];
+        }
+    } else {
+        /*
+         * The size of the table has changed or an element has been
+         * deleted. Must rehash, by inserting all old terms into the
+         * new (empty) table.
+         */
+        new_table->allocated = new_size;
+        new_table->num_entries = old_table->num_entries;
+        new_table->mask = new_size - 1;
+        for (i = 0; i < new_size; i++) {
+            new_table->term[i] = NIL;
+        }
+        for (i = 0; i < old_size; i++) {
+            if (is_tuple(old_table->term[i])) {
+                Eterm key = tuple_val(old_table->term[i])[1];
+                Uint entry_index = lookup(new_table, key);
+                ASSERT(is_nil(new_table->term[entry_index]));
+                new_table->term[entry_index] = old_table->term[i];
+            }
+        }
+    }
+    new_table->first_to_delete = 0;
+    new_table->num_to_delete = 0;
+    erts_atomic_init_nob(&new_table->refc, (erts_aint_t)1);
+    return new_table;
+}
+
+static void
+mark_for_deletion(HashTable* hash_table, Uint entry_index)
+{
+    hash_table->first_to_delete = entry_index;
+    hash_table->num_to_delete = 1;
+}
+
+static ErtsLiteralArea*
+term_to_area(Eterm tuple)
+{
+    ASSERT(is_tuple_arity(tuple, 2));
+    return (ErtsLiteralArea *) (((char *) tuple_val(tuple)) -
+                                offsetof(ErtsLiteralArea, start));
+}
+
+static void
+table_updater(void* data)
+{
+    HashTable* old_table;
+    HashTable* new_table;
+
+    old_table = (HashTable *) erts_atomic_read_nob(&the_hash_table);
+    new_table = (HashTable *) data;
+    ASSERT(new_table->num_to_delete == 0);
+    erts_atomic_set_nob(&the_hash_table, (erts_aint_t)new_table);
+    erts_schedule_thr_prgr_later_op(table_deleter,
```

Comment:
It seems safe to call `release_update_permission(1)` here directly after the_hash_table is set. `table_deleter` is a pure cleanup that no one needs to wait for.

Reply:
Fixed.